prusti-interface/src/specs/mod.rs

```diff
@@ -93,9 +101,66 @@ impl<'a, 'tcx> SpecCollector<'a, 'tcx> {
         self.determine_ghost_begin_ends(&mut def_spec);
         // TODO: remove spec functions (make sure none are duplicated or left over)
 
+        self.fetch_local_mirs(&mut def_spec);
+
+        if let Some(build_output_dir) = build_output_dir {
+            self.write_specs_to_file(&def_spec, &build_output_dir);
+            self.merge_specs_from_dependencies(&mut def_spec, &build_output_dir);
+        }
+
         def_spec
     }
 
+    fn get_local_crate_specs_filename(&self) -> String {
+        let name = (&self.tcx.sess().opts.crate_name).as_ref().unwrap();
+        let stable_id = &self.tcx.sess().local_stable_crate_id().to_u64();
+
+        return format!("{}-{:x}.bin", name, stable_id).to_owned();
+    }
+
+    fn get_serialized_mirs_dir(build_output_dir: &PathBuf) -> Box<PathBuf> {
+        let mut output_dir = Box::new(build_output_dir.clone());
+        output_dir.push("serialized_mirs");
```

Comment:
More descriptive name: `serialized_specs`? Or `serialized_prusti`?

Also I guess what is happening is we dump the serialised def spec map for each crate we compile into a `.bin` file in this directory. Is the TODO you mention [here](https://github.com/viperproject/prusti-dev/pull/1086#issuecomment-1182780570) the fact that we load all `.bin` files from there? If yes, then we should indeed fix this. For one, we get a directory that is never clean up from compiled crates that are no longer used, for two, we might keep loading broken/invalid specifications (though hopefully they would not be used…).

Reply:
Fixed, thanks! And yes, your understanding is correct.

I now pushed code that loads specifications only from crates referenced by `tcx.crates(())` (except the local crate). However, I am not sure if this is exactly the set of crates we want – do you know if there is a better way to list all dependency crates?

Reply:
That sounds like the correct query. I think "dependency crates" is a concept that exists in `cargo`, and not really in `rustc`. This should always include the set of crates we need, but might be larger, e.g. if `Cargo.toml` declares more dependencies, even if they are unused.

A more fine-grained spec import feature would be better still: we would have to locate the imported specs within `tcx.crates(())`, but at least we would only make use of the ones the user specifically requested.